Add GET routes for /user/tags and /user/tags/{id}

czan opened the pull request:
An operation that is useful for tools to be able to do is to list a user's tags without having to pull down the entire user object (which is likely significantly larger than the tags). This commit adds a route to GET /user/tags, to get a list of the user's tags, as well as a GET /user/tags/{id} to get the details of a particular tag.

This is relevant to issue #3621. No existing routes were changed, though, so the API is not completely restful.

czan committed:
Add GET routes for /user/tags and /user/tags/{id}

An operation that is useful for tools to be able to do is to list a
user's tags without having to pull down the entire user object (which is
likely significantly larger than the tags). This commit adds a route to
GET /user/tags, to get a list of the user's tags, as well as a GET
/user/tags/{id} to get the details of a particular tag.

This commit also includes tests for the two new routes.

crookedneighbor commented:
This reasonable to me, but currently it's blocked by a [new API test infrastructure we're building](https://github.com/HabitRPG/habitrpg/pull/6062). As soon that is merged in, we can add tests for this behavior and merge this in.

crookedneighbor commented:
The PR is merged in. Let me know if you need help writing tests.

czan commented:
I've rebased my changes and added tests for the GET routes (I was considering adding tests for the other /user/tags routes too, but decided I'd be better off keeping this pull request focused on the GET ones).

crookedneighbor commented on test/api/user/GET_user_tags_id.test.js:
```diff
@@ -0,0 +1,41 @@
+        expect(fetchedTag).to.eql(user.tags[0]);
+        done();
+      })
+      .catch((err) => {
```
This can be simplified to:

``` js
.catch(done);
```

crookedneighbor commented on test/api/user/GET_user_tags_id.test.js:
```diff
@@ -0,0 +1,41 @@
+        expect(err.code).to.eql(404);
+        done();
+      })
+      .catch((err) => {
```
This second catch block shouldn't be necessary.

czan replied:
I think we need both of these, actually. The first one checks that the error code is correct (because the `api.get(...)` call should return a promise which errors), then the second one catches the error from the `expect(...)` line (if there is one) and passes it to `done`.

czan replied:
I've thought more about this now, and I think what I have now is actually not the right way to do this. The `.then` step actually needs both a success and error handler at that point, not a separate `.then` followed by a `.catch`. I'll fix that up now.

crookedneighbor commented:
Overall, looks really good. Just a few things to clean up and it should be ready to merge.

czan commented:
Okay, I've replaced those unnecessary functions wrapping `done` and combined all my commits into the one concise commit for merging. Let me know if there are any more changes you want me to make!

czan commented:
Assuming the tests pass, this most recent commit should be ready to merge now. I've converted the tests to use promises instead of the `done` callback.

crookedneighbor commented:
The tests look great. Could you change their names to `GET-user_tags...` instead of `GET_user_tags...`, just to be consistent with the other tests?

czan commented:
Done! Sorry about that. I hadn't noticed the difference, but it's fixed up now.

crookedneighbor commented:
This will count towards a contributor tier! What's your User ID? You can find it in [Settings > API](https://habitrpg.com/#/options/settings/api). Be sure to only copy the User ID and _not_ your API Token.

czan commented:
Cheers!

User id: 3e3ef99b-3ee3-4114-a779-78332cb6e714

crookedneighbor commented:
Tier awarded! You can read more about [contributor tiers and rewards here](http://habitrpg.wikia.com/wiki/Contributor_Rewards).

Note that as your tier increases, it takes progressively more work to achieve the next one, so for the higher tiers, larger PRs or a few PRs are required. There's some information about that at Contributor Rewards, but feel free to ask here if you have questions!
